### PR TITLE
[expo-notification] Prevent from notification being dispatch twice

### PR DESCRIPTION
**Pull request by lukmccall**
# Why

Parts of #8135.

# How

Now, the user can specify which notification will be used on Android.
The scheme update: https://github.com/expo/universe/pull/5048

# Test Plan

- test-suite ✅
- using https://expo.io/notifications send notification and check where it will be dispatched

### CODE REVIEW

File: android/expoview/src/main/java/host/exp/exponent/fcm/ExpoFcmMessagingService.java [hunk 1]
```diff
@@ -31,7 +38,40 @@ public void onMessageReceived(RemoteMessage remoteMessage) {
...
+          onFailure();
+        }
+      }
+
+      @Override
+      public void onFailure() {
+        Log.w("expo-notifications", "Couldn't get experience from remote message.", null);
+        dispatchToNewNotificationModule(remoteMessage);
+        dispatchToOldNotificationModule(remoteMessage);
+      }
```
Comment (lukmccall) [inline android/expoview/src/main/java/host/exp/exponent/fcm/ExpoFcmMessagingService.java:66]: I'm not sure what should we do here 🤔

Comment (sjchmiela) [reply android/expoview/src/main/java/host/exp/exponent/fcm/ExpoFcmMessagingService.java:66]: Me too 🤔 Let's dispatch only to the old notifications module. 🤷 If no `useNewNotificationsAPI` falls back to using the old module, no manifest should fall back the same.

File: android/expoview/src/main/java/host/exp/exponent/fcm/ExpoFcmMessagingService.java [hunk 2]
```diff
@@ -31,7 +38,40 @@ public void onMessageReceived(RemoteMessage remoteMessage) {
       return;
     }
 
+    ExponentDB.experienceIdToExperience(remoteMessage.getData().get("experienceId"), new ExponentDB.ExperienceResultListener() {
```
Comment (sjchmiela) [inline android/expoview/src/main/java/host/exp/exponent/fcm/ExpoFcmMessagingService.java:41]: Does searching for it asynchronously work ok even when the app is terminated? What I'd be afraid of is starting new threads in a time-limited call-to-service. I believe Android catches it and doesn't terminate the app while the asynchronous callback is being executed, but who knows.

File: apps/test-suite/app.json [hunk 3]
```diff
@@ -8,6 +8,9 @@
     "platforms": ["android", "ios", "web"],
     "ios": {
       "bundleIdentifier": "io.expo.testsuite"
+    },
+    "android": {
+      "useNewNotificationsAPI": true
```
Comment (sjchmiela) [inline apps/test-suite/app.json:13]: Sorry for not finding this out earlier, I've looked at the code in `universe` and it looks like it's more common to write it like `Api` in camelcase

File: android/expoview/src/main/java/host/exp/exponent/fcm/ExpoFcmMessagingService.java [hunk 4]
```diff
@@ -31,7 +38,40 @@ public void onMessageReceived(RemoteMessage remoteMessage) {
...
+        dispatchToOldNotificationModule(remoteMessage);
+      }
+    });
+  }
+
+  private void dispatchToNewNotificationModule(RemoteMessage remoteMessage) {
     super.onMessageReceived(remoteMessage);
+  }
+
+  private void dispatchToOldNotificationModule(RemoteMessage remoteMessage) {
```
Comment (sjchmiela) [inline android/expoview/src/main/java/host/exp/exponent/fcm/ExpoFcmMessagingService.java:74]:
```suggestion
  private void dispatchToLegacyNotificationModule(RemoteMessage remoteMessage) {
```

File: android/expoview/src/main/java/host/exp/exponent/fcm/ExpoFcmMessagingService.java [hunk 5]
```diff
@@ -31,7 +38,40 @@ public void onMessageReceived(RemoteMessage remoteMessage) {
...
+      @Override
+      public void onFailure() {
+        Log.w("expo-notifications", "Couldn't get experience from remote message.", null);
+        dispatchToNewNotificationModule(remoteMessage);
+        dispatchToOldNotificationModule(remoteMessage);
+      }
+    });
+  }
+
+  private void dispatchToNewNotificationModule(RemoteMessage remoteMessage) {
```
Comment (sjchmiela) [inline android/expoview/src/main/java/host/exp/exponent/fcm/ExpoFcmMessagingService.java:70]:
```suggestion
  private void dispatchToNextNotificationModule(RemoteMessage remoteMessage) {
```

File: android/expoview/src/main/java/host/exp/exponent/fcm/ExpoFcmMessagingService.java [hunk 6]
```diff
@@ -31,7 +38,40 @@ public void onMessageReceived(RemoteMessage remoteMessage) {
...
     }
 
+    ExponentDB.experienceIdToExperience(remoteMessage.getData().get("experienceId"), new ExponentDB.ExperienceResultListener() {
+      @Override
+      public void onSuccess(ExperienceDBObject experience) {
+        try {
+          JSONObject manifest = new JSONObject(experience.manifest);
+          JSONObject androidSection = manifest.optJSONObject("android");
+          if (androidSection != null) {
+            boolean useNewNotificationsAPI = androidSection.optBoolean("useNewNotificationsAPI", false);
```
Comment (sjchmiela) [inline android/expoview/src/main/java/host/exp/exponent/fcm/ExpoFcmMessagingService.java:48]:
```suggestion
            boolean useNewNotificationsApi = androidSection.optBoolean("useNewNotificationsApi", false);
```

File: apps/test-suite/app.json (hunk 3 above)
Comment (sjchmiela) [reply apps/test-suite/app.json:13]: Hm, maybe `nextNotificationsApi` is even better? I don't know, what do you think @bbarthec?

File: android/expoview/src/main/java/host/exp/exponent/fcm/ExpoFcmMessagingService.java (hunk 2 above)
Comment (sjchmiela) [reply android/expoview/src/main/java/host/exp/exponent/fcm/ExpoFcmMessagingService.java:41]: Just to be on the safe side — do you think this work ok in standalone apps too? Should we add a note somewhere to ensure test this there?

File: android/expoview/src/main/java/host/exp/exponent/fcm/ExpoFcmMessagingService.java [hunk 7]
```diff
@@ -31,7 +38,40 @@ public void onMessageReceived(RemoteMessage remoteMessage) {
...
+            if (useNewNotificationsAPI) {
```
Comment (sjchmiela) [inline android/expoview/src/main/java/host/exp/exponent/fcm/ExpoFcmMessagingService.java:49]: Another conditional here would be sth like `|| (experience.sdkVersion >= 39 && < UNVERSIONED)`, right? Do you think we should add it already? 🤔

**Review by sjchmiela — COMMENTED** [review]
Great work!

File: android/expoview/src/main/java/host/exp/exponent/fcm/ExpoFcmMessagingService.java (hunk 2 above)
Comment (lukmccall) [reply android/expoview/src/main/java/host/exp/exponent/fcm/ExpoFcmMessagingService.java:41]: So, I've tested it and it seems to work even if the app was killed. But maybe, we should switch to sync the call 🤔

I believe that will work in standalone apps. However, we should test it :D

File: android/expoview/src/main/java/host/exp/exponent/fcm/ExpoFcmMessagingService.java [hunk 8]
```diff
@@ -31,7 +39,44 @@ public void onMessageReceived(RemoteMessage remoteMessage) {
       return;
     }
 
+    ExponentDB.experienceIdToExperience(remoteMessage.getData().get("experienceId"), new ExponentDB.ExperienceResultListener() {
+      @Override
+      public void onSuccess(ExperienceDBObject experience) {
+        try {
+          JSONObject manifest = new JSONObject(experience.manifest);
+          if (ABIVersion.toNumber(manifest.getString("sdkVersion")) >= 39) {
```
Comment (sjchmiela) [inline android/expoview/src/main/java/host/exp/exponent/fcm/ExpoFcmMessagingService.java:47]:
```suggestion
          // If an experience is on SDK newer than 38, that is SDK39 and beyond up till UNVERSIONED
          // we only use the new notifications API as it is going to be removed from SDK39.
          if (ABIVersion.toNumber(manifest.getString("sdkVersion")) >= 39) {
```

File: android/expoview/src/main/java/host/exp/exponent/fcm/ExpoFcmMessagingService.java [hunk 9]
```diff
@@ -31,7 +39,44 @@ public void onMessageReceived(RemoteMessage remoteMessage) {
...
+          }
+
+          JSONObject androidSection = manifest.optJSONObject("android");
+          if (androidSection != null) {
+            boolean useNextNotificationsApi = androidSection.optBoolean("useNextNotificationsApi", false);
+            if (useNextNotificationsApi) {
+              dispatchToNextNotificationModule(remoteMessage);
+              return;
+            }
+          }
```
Comment (sjchmiela) [inline android/expoview/src/main/java/host/exp/exponent/fcm/ExpoFcmMessagingService.java:59]: Could we scope this code to only run in SDK38? Someone in theory could push an update to an older-SDK-ed app with a custom field.

**Review by sjchmiela — APPROVED** [review]
Looks great!

File: android/expoview/src/main/java/host/exp/exponent/fcm/ExpoFcmMessagingService.java [hunk 10]
```diff
@@ -31,7 +39,51 @@ public void onMessageReceived(RemoteMessage remoteMessage) {
...
+            if (androidSection != null) {
+              boolean useNextNotificationsApi = androidSection.optBoolean("useNextNotificationsApi", false);
+              if (useNextNotificationsApi) {
+                dispatchToNextNotificationModule(remoteMessage);
+                return;
+              }
+            }
+          }
+
+          dispatchToLegacyNotificationModule(remoteMessage);
```
Comment (sjchmiela) [inline android/expoview/src/main/java/host/exp/exponent/fcm/ExpoFcmMessagingService.java:67]:
```suggestion
          if (sdkVersion >= 39) {
            dispatchToNextNotificationModule(remoteMessage);
          } else if (sdkVersion == 38) {
            // In SDK38 we want to let people decide which notifications API to use,
            // the next or the legacy one.
            JSONObject androidSection = manifest.optJSONObject("android");
            if (androidSection != null) {
              boolean useNextNotificationsApi = androidSection.optBoolean("useNextNotificationsApi", false);
              if (useNextNotificationsApi) {
                dispatchToNextNotificationModule(remoteMessage);
                return;
              }
            }
          } else {
            // If it's an older experience let's only use legacy notifications API
            dispatchToLegacyNotificationModule(remoteMessage);
          }
```

**Review by sjchmiela — APPROVED** [review]
**cyk**!

**Review by sjchmiela — APPROVED** [review]
🥳🥳🥳

**Pull request merged**